gensim/models/doc2vec.py

```diff
@@ -91,265 +90,12 @@
-        doctag_locks : list of float, optional
-            The lock factors for each tag.
-
-        Returns
-        -------
-        int
-            Number of words in the input document that were actually used for training (they were found in the
-            vocabulary and they were not discarded by negative sampling).
-
-        """
-        if word_vectors is None:
-            word_vectors = model.wv.syn0
-        if word_locks is None:
-            word_locks = model.syn0_lockf
-        if doctag_vectors is None:
-            doctag_vectors = model.docvecs.doctag_syn0
-        if doctag_locks is None:
-            doctag_locks = model.docvecs.doctag_syn0_lockf
-
-        word_vocabs = [model.wv.vocab[w] for w in doc_words if w in model.wv.vocab
-                       and model.wv.vocab[w].sample_int > model.random.rand() * 2 ** 32]
-        doctag_len = len(doctag_indexes)
-        if doctag_len != model.dm_tag_count:
-            return 0  # skip doc without expected number of doctag(s) (TODO: warn/pad?)
-
-        null_word = model.wv.vocab['\0']
-        pre_pad_count = model.window
-        post_pad_count = model.window
-        padded_document_indexes = (
-                (pre_pad_count * [null_word.index])  # pre-padding
-                + [word.index for word in word_vocabs if word is not None]  # elide out-of-Vocabulary words
-                + (post_pad_count * [null_word.index])  # post-padding
-        )
-
-        for pos in range(pre_pad_count, len(padded_document_indexes) - post_pad_count):
-            word_context_indexes = (
-                    padded_document_indexes[(pos - pre_pad_count): pos]  # preceding words
-                    + padded_document_indexes[(pos + 1):(pos + 1 + post_pad_count)]  # following words
-            )
-            predict_word = model.wv.vocab[model.wv.index2word[padded_document_indexes[pos]]]
-            # numpy advanced-indexing copies; concatenate, flatten to 1d
-            l1 = concatenate((doctag_vectors[doctag_indexes], word_vectors[word_context_indexes])).ravel()
-            neu1e = train_cbow_pair(model, predict_word, None, l1, alpha,
-                                    learn_hidden=learn_hidden, learn_vectors=False)
-
-            # filter by locks and shape for addition to source vectors
-            e_locks = concatenate((doctag_locks[doctag_indexes], word_locks[word_context_indexes]))
-            neu1e_r = (neu1e.reshape(-1, model.vector_size)
-                       * np_repeat(e_locks, model.vector_size).reshape(-1, model.vector_size))
-
-            if learn_doctags:
-                np_add.at(doctag_vectors, doctag_indexes, neu1e_r[:doctag_len])
-            if learn_words:
-                np_add.at(word_vectors, word_context_indexes, neu1e_r[doctag_len:])
-
-        return len(padded_document_indexes) - pre_pad_count - post_pad_count
+    #
+    # We could raise utils.NO_CYTHON right here, but let's delay it until
```

Comment:
you no need to catch/raise anything here if extension available, if not - the user should use wheel / compile self.